src/win/fs.c

```diff
@@ -788,123 +786,205 @@ void fs__mkdtemp(uv_fs_t* req) {
+   * It's important that this buffer can hold at least one entry, regardless
+   * of the length of the file names present in the enumerated directory.
+   * A file name is at most 256 WCHARs long.
+   * According to MSDN, the buffer must be aligned at an 8-byte boundary.
+   */
+  __declspec(align(8)) char buffer[8192];
+
+  STATIC_ASSERT(sizeof buffer >=
+                sizeof(FILE_DIRECTORY_INFORMATION) + 256 * sizeof(WCHAR));
+
+  /* Open the directory. */
+  dir_handle =
+      CreateFileW(req->pathw,
+                  FILE_LIST_DIRECTORY | SYNCHRONIZE,
+                  FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+                  NULL,
+                  OPEN_EXISTING,
+                  FILE_FLAG_BACKUP_SEMANTICS,
+                  NULL);
+  if (dir_handle == INVALID_HANDLE_VALUE)
+    goto win32_error;
+
+  /* Read the first chunk. */
+  status = pNtQueryDirectoryFile(dir_handle,
+                                 NULL,
+                                 NULL,
+                                 NULL,
+                                 &iosb,
+                                 &buffer,
+                                 sizeof buffer,
+                                 FileDirectoryInformation,
+                                 FALSE,
+                                 NULL,
+                                 TRUE);
+
+  /* If the handle is not a directory, we'll get STATUS_INVALID_PARAMETER.
+   * This should be reported back as UV_ENOTDIR.
+   */
+  if (status == STATUS_INVALID_PARAMETER)
+    goto not_a_directory_error;
+
+  while (NT_SUCCESS(status)) {
+    char* position = buffer;
+    size_t next_entry_offset = 0;
+
+    do {
+      FILE_DIRECTORY_INFORMATION* info;
+      uv__dirent_t* dirent;
+
+      size_t wchar_len;
+      size_t utf8_len;
+
+      /* Obtain a pointer to the current directory entry. */
+      position += next_entry_offset;
+      info = (FILE_DIRECTORY_INFORMATION*) position;
+
+      /* Fetch the offset to the next directory entry. */
+      next_entry_offset = info->NextEntryOffset;
+
+      /* Compute the length of the filename in WCHARs. */
+      wchar_len = info->FileNameLength / sizeof(WCHAR);
+
+      /* Skip over '.' and '..' entries. */
+      if (wchar_len == 1 && info->FileName[0] == L'.')
+        continue;
+      if (wchar_len == 2 && info->FileName[0] == L'.' &&
+          info->FileName[1] == L'.')
+        continue;
+
+      /* Compute the space required to store the filename as UTF-8. */
+      utf8_len = WideCharToMultiByte(
+          CP_UTF8, 0, &info->FileName[0], wchar_len, NULL, 0, NULL, NULL);
+      if (utf8_len == 0)
+        goto win32_error;
+
+      /* Resize the dirent array if needed. */
+      if (dirents_used >= dirents_size) {
+        size_t new_dirents_size =
+            dirents_size == 0 ? DIRENTS_INITIAL_SIZE : dirents_size << 1;
+        uv__dirent_t** new_dirents = (uv__dirent_t**) realloc(
```

Comment:
I don't mind, actually I didn't know MSVC generated a warning about it :-O
On Jan 4, 2015 7:02 PM, "Bert Belder" <notifications@github.com> wrote:

> In src/win/fs.c
> <https://github.com/libuv/libuv/pull/105#discussion-diff-22441096>:
>
> > +        continue;
> > +      if (wchar_len == 2 && info->FileName[0] == L'.' &&
> > +          info->FileName[1] == L'.')
> > +        continue;
> > +
> > +      /* Compute the space required to store the filename as UTF-8. */
> > +      utf8_len = WideCharToMultiByte(
> > +          CP_UTF8, 0, &info->FileName[0], wchar_len, NULL, 0, NULL, NULL);
> > +      if (utf8_len == 0)
> > +        goto win32_error;
> > +
> > +      /* Resize the dirent array if needed. */
> > +      if (dirents_used >= dirents_size) {
> > +        size_t new_dirents_size =
> > +            dirents_size == 0 ? DIRENTS_INITIAL_SIZE : dirents_size << 1;
> > +        uv__dirent_t** new_dirents = (uv__dirent_t**) realloc(
>
> Do you mind keeping it? It avoids msvc warnings.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/libuv/libuv/pull/105/files#r22441096>.
>